src/expr/symbol_table.cpp

```diff
@@ -255,18 +314,121 @@ void SymbolTable::Implementation::reset() {
+    if((*it).second!=obj) {
+      if(!isOverloadedFunction((*it).second)) {
+        // mark previous as overloaded
+        markOverloaded(name, (*it).second);
+      }
+      // mark this as overloaded
+      markOverloaded(name, obj);
+    }
+  }
+}
+
+void SymbolTable::Implementation::markOverloaded(const string& name, Expr obj) {
+  Trace("parser-overloading") << "Overloaded function : " << name;
+  Trace("parser-overloading") << " with type " << obj.getType() << std::endl;
+  d_overloaded_symbols->insert(obj);
+  // get the argument types
+  Type t = obj.getType();
+  Type rangeType = t;
+  std::vector< Type > argTypes;
+  if(t.isFunction()) {
+    argTypes = ((FunctionType)t).getArgTypes();
+    rangeType = ((FunctionType)t).getRangeType();
+  }else if(t.isConstructor()) {
+    argTypes = ((ConstructorType)t).getArgTypes();
+    rangeType = ((ConstructorType)t).getRangeType();
+  }else if(t.isTester()) {
+    argTypes.push_back( ((TesterType)t).getDomain() );
+    rangeType = ((TesterType)t).getRangeType();
+  }else if(t.isSelector()) {
+    argTypes.push_back( ((SelectorType)t).getDomain() );
+    rangeType = ((SelectorType)t).getRangeType();
+  }
+  // add to the trie
+  TypeArgTrie * tat = &d_overload_type_arg_trie[name];
+  for(unsigned i=0; i<argTypes.size(); i++) {
+    tat = &(tat->d_children[argTypes[i]]);
+  }
+  tat->d_symbols[rangeType] = obj;
+}
+  
+bool SymbolTable::Implementation::isOverloadedFunction(Expr fun) const {
+  return d_overloaded_symbols->find(fun)!=d_overloaded_symbols->end();
+}
+
+Expr SymbolTable::Implementation::getOverloadedConstantForType(const std::string& name, Type t) const {
+  std::unordered_map< std::string, TypeArgTrie >::const_iterator it = d_overload_type_arg_trie.find(name);
+  if(it!=d_overload_type_arg_trie.end()) {
+    std::map< Type, Expr >::const_iterator its = it->second.d_symbols.find(t);
+    if(its!=it->second.d_symbols.end()) {
+      return its->second;
+    }
+  }
+  return d_nullExpr;
+}
+
+Expr SymbolTable::Implementation::getOverloadedFunctionForTypes(const std::string& name, 
+                                                                const std::vector< Type >& argTypes) const {
+  std::unordered_map< std::string, TypeArgTrie >::const_iterator it = d_overload_type_arg_trie.find(name);
+  if(it!=d_overload_type_arg_trie.end()) {
+    const TypeArgTrie * tat = &it->second;
+    for(unsigned i=0; i<argTypes.size(); i++) {
+      std::map< Type, TypeArgTrie >::const_iterator itc = tat->d_children.find(argTypes[i]);
+      if(itc!=tat->d_children.end()) {
+        tat = &itc->second;
+      }else{
+        // no functions match 
+        return d_nullExpr;
+      }
+    }
+    // now, we must ensure that there is one and only one active symbol at this node
+    Expr retExpr;
+    for(std::map< Type, Expr >::const_iterator its = tat->d_symbols.begin(); its != tat->d_symbols.end(); ++its) {
+      Expr expr = its->second;
+      if(isOverloadedFunction(expr)) {
```

Comment:
Why this check? Is it because the trie is not context-dependent?

Reply:
Yes, the trie is context-independent so this is the check that makes it "virtually context-dependent".

I do believe this is an important coding style to use in CVC4 (i.e. only simple things are context-dependent and the more complicated data structures are context-independent), hence I would argue for keep the trie context-independent, to address your previous comment.  Perhaps a solution here is to encapsulate the trie + context dependent set as its own class.

Reply:
I went ahead and refactored so there is no confusion regarding this point, there is now a OverloadTypeTrie class that should be interpreted as context-dependent.